File: bfs2.cpp
#include <bits/stdc++.h>
using namespace std;

#define ROW 9
#define COL 10

struct Point
{
  int x;
  int y;
};

struct Node
{
  Point pt;
  int dist;
};

bool isValid(int row, int col)
{
  return (row >= 0) && (row < ROW) && (col >= 0) && (col < COL);
}

int rowNum[] = {-1, 0, 0, 1};
int colNum[] = {0, -1, 1, 0};

void printPath(vector<Node> &path)
{
  for (int i = 0; i < path.size(); i++)
    cout << " (" << path[i].pt.x << "," << path[i].pt.y << "," << path[i].dist << ") ";
  cout << endl;       
}

bool isNotVisited(int row, int col, vector<Node> &path)
{
  for (int i = 0; i < path.size(); i++)
  {
    if (path[i].pt.x == row && path[i].pt.y == col)
      return false;
  }
  return true;
}

void BFS(int mat[][COL], Point src, Point dest)
{
  if (!mat[src.x][src.y] || !mat[dest.x][dest.y])
  {
    cout << "No path\n";
    return;
  }

  queue<vector<Node>> q;
  vector<Node> path;
  path.push_back({src, 0});
  q.push(path);

  while (!q.empty())
  {
    path = q.front();
    q.pop();

    Node last = path[path.size() - 1];
    cout << last.pt.x << " " << last.pt.y << endl;

    if (last.pt.x == dest.x && last.pt.y == dest.y)
    {
      printPath(path);
      return;
    }

    for (int i = 0; i < 4; i++)
    {
      int row = last.pt.x + rowNum[i];
      int col = last.pt.y + colNum[i];

      if (isValid(row, col) && isNotVisited(row, col, path))
      {
        vector<Node> newPath(path);
        newPath.push_back({row, col, last.dist + 1});
        q.push(newPath);
      }
    }
  }
}

int main()
{
  int mat[ROW][COL] =
      {
          {1, 0, 1, 1, 1, 1, 0, 1, 1, 1},
          {1, 0, 1, 0, 1, 1, 1, 0, 1, 1},
          {1, 1, 1, 0, 1, 1, 0, 1, 0, 1},
          {0, 0, 0, 0, 1, 0, 0, 0, 0, 1},
          {1, 1, 1, 0, 1, 1, 1, 0, 1, 0},
          {1, 0, 1, 1, 1, 1, 0, 1, 0, 0},
          {1, 0, 0, 0, 0, 0, 0, 0, 0, 1},
          {1, 0, 1, 1, 1, 1, 0, 1, 1, 1},
          {1, 1, 0, 0, 0, 0, 1, 0, 0, 1}};
  Point src = {0, 0};
  Point dest = {3, 4};

  BFS(mat, src, dest);
  //cout << isValid(0, 0) << endl;
}

File: bfs2.js
function Point(x, y) {
    this.x = x;
    this.y = y;
}

function Node(pt, dist) {
    this.pt = pt;
    this.dist = dist;
}

var d = [
    { x: 0, y: 1 },
    { x: 1, y: 0 },
    { x: 0, y: -1 },
    { x: -1, y: 0 }
];

function printPath(path) {

    console.log("Final path: ");
    console.log(path);
}

function isNotVisited(row, col, path) {
    for (let i = 0; i < path.length; i++) {
        if (path[i].pt.x === row && path[i].pt.y === col)
            return false;
    }
    return true;
}

function isValid(maze, width, height, row, col) {
    return (row >= 0) && (row < width) && (col >= 0) && (col < height) && maze[row][col] == 1; //&& !visited[row][col];
}

function equal(newPath, path) {
    for (var i = 0; i < path.length; i++)
        newPath.push(path[i]);
}

function bfsUtil(queue, width, height) {
    let path = queue.pop();
    let last = path[path.length - 1];

    if (last.pt.x === dest.x && last.pt.y === dest.y) {
        pathFound = true;
        printPath(path);
        return;
    }

    for (let i = 0; i < 4; i++) {
        let row = last.pt.x + d[i].x;
        let col = last.pt.y + d[i].y;

        if (isValid(maze, width, height, row, col) && isNotVisited(row, col, path)) {
            //visited[row][col] = true
            let newPath = [];
            equal(newPath, path);
            let node = new Node(new Point(row, col), last.dist + 1);
            newPath.push(node);
            queue.push(newPath);
        }
    }

}

function bfs(maze, src, dest) {
    let pathFound = false;

    if (maze[src.x][src.y] || maze[dest.x][dest.y]) {

        let queue = [];
        let path = [];

        let height = maze.length;
        let width = maze[0].length;

        visited = [];
        for (var i = 0; i < height; i++) {
            visited.push([]);
            for (var j = 0; j < width; j++) {
                visited[i].push(false);
            }
        }

        let s = new Node(src, 0);
        path.push(s);
        queue.push(path);
        visited[src.x][src.y] = true;


        while (queue.length > 0) {

        }
    }
    if (!pathFound)
        console.log("No path found");
}



var maze = [
    [1, 1, 1, 1, 1, 0, 0, 1, 1, 1],
    [0, 1, 1, 1, 1, 1, 0, 1, 0, 1],
    [0, 0, 1, 0, 1, 1, 1, 0, 0, 1],
    [1, 0, 1, 1, 1, 0, 1, 1, 0, 1],
    [0, 0, 0, 1, 0, 0, 0, 1, 0, 1],
    [1, 0, 1, 1, 1, 0, 0, 1, 1, 0],
    [0, 0, 0, 0, 1, 0, 0, 1, 0, 1],
    [0, 1, 1, 1, 1, 1, 1, 1, 0, 0],
    [1, 1, 1, 1, 1, 0, 0, 1, 1, 1],
    [0, 0, 1, 0, 0, 1, 1, 0, 0, 1]
];

var source = new Point(0, 0);
var dest = new Point(5, 8);
bfs(maze, source, dest);